core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java

```diff
@@ -557,135 +566,176 @@ public void applyState(final Address sender, int topologyId, boolean pushTransfe
          }
       }
       if (inboundTransfer != null) {
-         if (stateChunk.getCacheEntries() != null) {
-            doApplyState(sender, stateChunk.getSegmentId(), stateChunk.getCacheEntries());
-         }
-
-         inboundTransfer.onStateReceived(stateChunk.getSegmentId(), stateChunk.isLastChunk());
+         return doApplyState(sender, stateChunk.getSegmentId(), stateChunk.getCacheEntries())
+                   .thenAccept(
+                      v -> inboundTransfer.onStateReceived(stateChunk.getSegmentId(), stateChunk.isLastChunk()));
       } else {
          if (cache.wired().getStatus().allowInvocations()) {
             log.ignoringUnsolicitedState(sender, stateChunk.getSegmentId(), cacheName);
          }
       }
+      return CompletableFutures.completedNull();
    }
 
-   private void doApplyState(Address sender, int segmentId, Collection<InternalCacheEntry> cacheEntries) {
+   private CompletionStage<?> doApplyState(Address sender, int segmentId,
+                                           Collection<InternalCacheEntry> cacheEntries) {
+      if (cacheEntries == null || cacheEntries.isEmpty())
+         return CompletableFutures.completedNull();
+
       if (trace) log.tracef("Applying new state chunk for segment %d of cache %s from node %s: received %d cache entries",
-            segmentId, cacheName, sender, cacheEntries.size());
+                            segmentId, cacheName, sender, cacheEntries.size());
 
       // CACHE_MODE_LOCAL avoids handling by StateTransferInterceptor and any potential locks in StateTransferLock
       boolean transactional = transactionManager != null;
-      for (InternalCacheEntry e : cacheEntries) {
+      if (transactional) {
+         Object key = "N/A";
+         Transaction transaction = new ApplyStateTransaction();
+         InvocationContext ctx = icf.createInvocationContext(transaction, false);
+         LocalTransaction localTransaction = ((LocalTxInvocationContext) ctx).getCacheTransaction();
          try {
-            InvocationContext ctx;
-            if (transactional) {
-               transactionManager.begin();
-               ctx = icf.createInvocationContext(transactionManager.getTransaction(), true);
-               ((TxInvocationContext) ctx).getCacheTransaction().setStateTransferFlag(PUT_FOR_STATE_TRANSFER);
+            localTransaction.setStateTransferFlag(PUT_FOR_STATE_TRANSFER);
+            for (InternalCacheEntry e : cacheEntries) {
+               // CallInterceptor will preserve the timestamps if the metadata is an InternalMetadataImpl instance
+               key = e.getKey();
+               CompletableFuture<?> future = invokePut(segmentId, ctx, e);
+               if (!future.isDone()) {
+                  throw new IllegalStateException("State transfer in-tx put should always be synchronous");
+               }
+            }
+         } catch (Throwable t) {
+            logApplyException(t, key);
+            if (localTransaction != null) {
```

Comment:
Can this be `null`? Shouldn't line 656 guarantee it is always non null?

Reply:
Yeah, it's a leftover from the time I was still using `tm.begin()`. Weird that IntelliJ didn't offer to simplify the if for me...